### DISCUSSION TITLE
Attempting to save data to a note

### DISCUSSION THREAD

**Discussion by sottey** [TriliumNext/Trilium#4217] [Q&A] [answered]
I have another strange thing happening...

I have a note which contains some json data. I have successfully (with help from some great people here) loaded this data into my FrontendJS app I am writing.

For changes that are made, I need to then take that changed JSON and save it back to the original note. However, something very strange is happening.

Here is the function that is trying to save the data:
```
async function setNoteContent(noteID, content) {
        await api.runOnBackend((noteID, content) => {
            const targetNote = api.getNote(noteID);
            targetNote.setContent(content, null);
         });
                
        await api.waitUntilSynced();
}
```

The note id passed in is ln7hIzdFYEVh. However, when I attempt to save, I am getting this error:
```
desktop.js:2 Uncaught (in promise) Error: server error: Load of script note "JS (Trilium)" (zxM7mee50qyV) failed with: Cannot set null content to note 'hK6Czvugwyhl'
    at v.runOnBackend (desktop.js:2:80548)
    at async setNoteContent (eval at <anonymous> (desktop.js:2:35713), <anonymous>:158:9)
```

hK6Czvugwyhl is a completely different note. It is a note I deleted this afternoon. I did an integrity fix, a consistency check and a DB vacuum and now I get the same error but with ID 66nQiUxkVkJx. That is the ID of the SQL Console I had just used to find out what the previous ID was.

My confusion is this... It seems that an id gets "stuck" and when I call getContents with an ID, it is ignoring that and using a cached ID.

Anyone else seen this?

I am using the stable release (60.4) on an Ubuntu server running the Trilium Docker instance. I have https all set and working, but no sync enabled (only using the web interface).

Thanks in advance,
Sean

**Comment by zerebos** [comment, marked as answer]
`runOnBackend` takes two arguments, the function to run, and the `arguments` array to pass to said function. You need to pass your arguments in so they can be serialized over to the backend context.

Try doing something like this:
```js
async function setNoteContent(noteID, content) {
        await api.runOnBackend((id, newContent) => {
            const targetNote = api.getNote(id);
            targetNote.setContent(newContent, null);
         }, [noteID, content]);
                
        await api.waitUntilSynced();
}
```

Think of the function you pass as being stringified (because it is), so the wrapping function's parameters are not actually in scope when the function runs.

Edit: Updated the variables to be different to show that they are in fact not connected at all.

**Reply by sottey** [reply]
That worked like a charm! Thank you!

I appreciate the help getting me up to speed, even if it wasn't Trilium, but my lack of JS knowledge. You all are the best!

**Reply by zerebos** [reply]
No problem, happy to help! And don't worry, even seasoned JS devs can mess this sort of thing up because they aren't used to dealing with multiple contexts like this. You can only really experience this sort of thing in an electron (or similar) based app.

**Reply by zadam** [reply]
`api.runOnBackend()` is a bit magical (as @rauenzi explained) in how it switches between two different JS processes. But there's a point where every magic breaks ...
